src/gen/conditions.rs

```diff
@@ -0,0 +1,1643 @@
+        }
+        ASSERT_COIN_ANNOUNCEMENT => {
+            let id = sanitize_hash(a, first(a, c)?, 32, ErrorCode::AssertCoinAnnouncementFailed)?;
+            Ok(Condition::AssertCoinAnnouncement(id))
+        }
+        CREATE_PUZZLE_ANNOUNCEMENT => {
+            let msg = sanitize_announce_msg(a, first(a, c)?, ErrorCode::InvalidPuzzleAnnouncement)?;
+            Ok(Condition::CreatePuzzleAnnouncement(msg))
+        }
+        ASSERT_PUZZLE_ANNOUNCEMENT => {
+            let id = sanitize_hash(
+                a,
+                first(a, c)?,
+                32,
+                ErrorCode::AssertPuzzleAnnouncementFailed,
+            )?;
+            Ok(Condition::AssertPuzzleAnnouncement(id))
+        }
+        ASSERT_MY_COIN_ID => {
+            let id = sanitize_hash(a, first(a, c)?, 32, ErrorCode::AssertMyCoinIdFailed)?;
+            Ok(Condition::AssertMyCoinId(id))
+        }
+        ASSERT_MY_PARENT_ID => {
+            let id = sanitize_hash(a, first(a, c)?, 32, ErrorCode::AssertMyParentIdFailed)?;
+            Ok(Condition::AssertMyParentId(id))
+        }
+        ASSERT_MY_PUZZLEHASH => {
+            let id = sanitize_hash(a, first(a, c)?, 32, ErrorCode::AssertMyPuzzlehashFailed)?;
+            Ok(Condition::AssertMyPuzzlehash(id))
+        }
+        ASSERT_MY_AMOUNT => {
+            let amount = parse_amount(
+                a,
+                first(a, c)?,
+                ErrorCode::AssertMyAmountFailed,
+                range_cache,
+                flags,
+            )?;
+            Ok(Condition::AssertMyAmount(amount))
+        }
+        ASSERT_SECONDS_RELATIVE => Ok(Condition::AssertSecondsRelative(parse_seconds(
+            a,
+            first(a, c)?,
+            ErrorCode::AssertSecondsRelative,
+            range_cache,
+            flags,
+        )?)),
+        ASSERT_SECONDS_ABSOLUTE => Ok(Condition::AssertSecondsAbsolute(parse_seconds(
+            a,
+            first(a, c)?,
+            ErrorCode::AssertSecondsAbsolute,
+            range_cache,
+            flags,
+        )?)),
+        ASSERT_HEIGHT_RELATIVE => Ok(Condition::AssertHeightRelative(parse_height(
+            a,
+            first(a, c)?,
+            ErrorCode::AssertHeightRelative,
+            range_cache,
+            flags,
+        )?)),
+        ASSERT_HEIGHT_ABSOLUTE => Ok(Condition::AssertHeightAbsolute(parse_height(
+            a,
+            first(a, c)?,
+            ErrorCode::AssertHeightAbsolute,
+            range_cache,
+            flags,
+        )?)),
+        _ => Err(ValidationErr(c, ErrorCode::InvalidConditionOpcode)),
+    }
+}
+
+#[derive(Debug)]
+pub struct SpendConditionSummary {
+    pub coin_id: Arc<[u8; 32]>,
+    pub puzzle_hash: NodePtr,
+    // conditions
+    // The sum of all reserve fee conditions
+    pub reserve_fee: u64,
+    // the highest height/time conditions (i.e. most strict)
+    pub height_relative: u32,
+    pub height_absolute: u32,
+    pub seconds_relative: u64,
```

Comment:
I'm still very scared about the implicit `(ASSERT_SECONDS_AGE_EXCEEDS 0)` this code generates. How hard would it be to change `seconds_relative` to `Option<u64>`? In particular, if that assert can ever fail, this is a soft fork (and it could fail if block times are not guaranteed to be monotonic, or if the "exceeds" condition is `>` rather than `>=`).

Similar concerns for `height_relative`.

Reply:
There has been discussion in `chia_network#consensus` about this, but I'm not 100% convinced. I would prefer @mariano54 weigh in also, or research the two questions yourself: (1) is "exceeds" non-strict, ie. >= 0 is okay; and (2) are blocks monotonic.

Reply:
This seems to suggest transaction block timestamps are purely monotonic increasing https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/block_header_validation.py#L822

Reply:
This seems to suggest we're okay with `>` vs `>=` https://github.com/Chia-Network/chia-blockchain/blob/7d55c39ab18609f47ce90106314244f445496947/chia/full_node/mempool_check_conditions.py#L98

Reply:
This seems to suggest we're okay with `>` vs `>=` for block height https://github.com/Chia-Network/chia-blockchain/blob/7d55c39ab18609f47ce90106314244f445496947/chia/full_node/mempool_check_conditions.py#L66

Reply:
I'm still nervous and would prefer is someone else also review this.

Reply:
there is no condition created for 0. 0 is assumed to always be true and is ignored.
See the places where these conditions are turned into conditions on their way to python:

https://github.com/Chia-Network/clvm_rs/pull/95/files#diff-2c8dc65d12ea32516876ce48b208f38ea2107a7f5a4f74d966a297a165cf345fR139

Reply:
if the assumption is wrong, that `ASSERT_TIME_RELATIVE 0` isn't always true. The case that would fail would be:

* a coin is created with an explicit `ASSERT_TIME_RELATIVE 0`
* in the same block, that coin is being spent

With my patch, that would be OK, the condition would never be checked.
With the current code, I believe it's also ok. but [this check](https://github.com/Chia-Network/chia-blockchain/blob/7d55c39ab18609f47ce90106314244f445496947/chia/full_node/mempool_check_conditions.py#L111) *would* fail if `expected` and `timestamp` are the same.

Reply:
@Yostra as far as I can tell, my example above *would* fail currently. Would you mind taking a look as well?

* The timestamp of the ephemeral coin is `block.foliage_transaction_block.timestamp` (see [here](https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/block_body_validation.py#L375)).
* The timestamp passed in to `mempool_check_conditions_dict` is the same, `block.foliage_transaction_block.timestamp` (see [here](https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/block_body_validation.py#L465)).
* The conditions checker checks current-time < expected (not <=). See [here](https://github.com/Chia-Network/chia-blockchain/blob/main/chia/full_node/mempool_check_conditions.py#L111)

If you have an ephemeral coin with a condition `ASSERT_SECONDS_RELATIVE 0`, it would fail, because the check for that condition is that the current timestamp is *less* than the expected timestamp for the block (which is the same for ephemeral coins) + the relative time, which is 0. i.e. you end up testing `current_time < current_time`, which is false.

i.e. `ASSERT_SECONDS_RELATIVE 0` is not unconditionally true.

Reply:
@richardkiss do you agree that it's *not* safe to assume `ASSERT_SECONDS_RELATIVE 0) to be true unconditionally? However, `ASSERT_SECONDS_RELATIVE -1` *is*

Reply:
the same goes for `ASSERT_HEIGHT_RELATIVE`. 0 means it cannot be spent in the same block (ifaict)

Reply:
I see. Rather than use 0 as the indicator to not create a condition, you might consider using `None`... any reason to not do that?

I'm reasonably convinced it the way you have it not would not change behaviour. But, it still feels a bit icky. Using `None` rather than `0` just seems a bit more explicit.

Reply:
the main reason is that `Option<uint64>` is more expensive than just `u64`. but it seems 0 is a valid condition that affects ephemeral coins, so I will need to change this. I just need a test first

Reply:
this has been changed for `height_relative`, to use `Option<u32>`. I expect the `Option` to double the size of the field, but 0 is relevant for relative height